src/main/java/org/opensearch/ad/feature/ScriptMaker.java

```diff
@@ -0,0 +1,118 @@
+     * the same document.  Scripts come to the rescue: With a script to retrieve terms
+     * from multiple fields, we can still use terms aggregation to partition data.
+     * The script disables the global ordinals optimization and will be slower than
+     * collecting terms from a single field.  Still, it gives us the flexibility to
+     * implement this option at search time. For a simple example, consider the
+     * following query about the number field’s sum aggregation on buckets partitioned
+     * by category_field_1 and category_field_2 from index test.
+     *
+     * Query:
+     * GET /test/_search
+        {
+            "aggregations": {
+                "term_agg": {
+                    "terms": {
+                        "script": {
+                            "source": "['category_field_1': doc['category_field_1'].value,
+                             'category_field_2': doc['category_field_2'].value]",
+                            "lang": "painless"
+                        }
+                    },
+                    "aggregations": {
+                        "sum_number": {
+                            "sum": {
+                                "field": "number"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+       *
+       * Result:
+       *"aggregations": {
+            "term_agg": {
+                "doc_count_error_upper_bound": 0,
+                "sum_other_doc_count": 0,
+                "buckets": [
+                    {
+                        "key": "{category_field_1=app_0, category_field_2=server_1}",
+                        "doc_count": 1,
+                        "sum_number": {
+                            "value": 1449.0
+                        }
+                    },
+                    {
+                        "key": "{category_field_1=app_1, category_field_2=server_1}",
+                        "doc_count": 1,
+                        "sum_number": {
+                            "value": 5200.0
+                        }
+                    },
+              ...
+       *
+       * I put two categorical field in a map for parsing the results.  Otherwise,
+       * I won't know which categorical value is for which field.
+     * @param fields categorical fields
+     * @return script to use in terms aggregation
+     */
+    public static Script makeTermsScript(List<String> fields) {
+        StringBuffer format = new StringBuffer();
+        // in painless, a map is sth like [a:b, c:d]
+        format.append("[");
+        for (int i = 0; i < fields.size(); i++) {
+            if (i > 0) {
+                format.append(",");
+            }
+            format.append("\'");
+            format.append(fields.get(i));
+            format.append("\':");
+            format.append(String.format(Locale.ROOT, template, fields.get(i)));
```

Comment:
Rather than handling only part of this via string template, can you change to handle the entire mapping as a template? Example template: `"\'%s\': doc['%s'].value"`

Reply:
yeah, done